impeller/core/host_buffer.cc

```diff
@@ -56,6 +56,19 @@ BufferView HostBuffer::Emplace(const void* buffer, size_t length) {
   return BufferView{shared_from_this(), GetBuffer(), Range{old_length, length}};
 }
 
+BufferView HostBuffer::EmplaceCallback(size_t length,
+                                       size_t align,
+                                       const EmplaceProc& cb) {
+  auto old_length = GetLength();
+  if (!Truncate(old_length + length)) {
+    return {};
+  }
+  generation_++;
+  cb(GetBuffer() + old_length);
```

Comment:
Emplace and EmplaceCallback can be dried up by making Emplace call EmplaceCallback with a callback that does the `memmove`.

Reply:
Also, let's check that CB is valid. If not, emplacing undefined data is fine.

Reply:
Alternatively consider making this return a `std::pair`

Reply:
e.g. `std::pair<BufferView, uintptr_t> Reserve(size_t length, size_t align);` and then on this line you 

```cpp
return std::make_pair(BufferView{...}, GetBuffer() + old_length);
```

Reply:
Emplace could still be dried up as Chinmay is suggesting to internally call `Reserve` and use the pointer.